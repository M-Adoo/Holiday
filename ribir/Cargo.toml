[package]
authors.workspace = true
categories.workspace = true
description.workspace = true
documentation.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "ribir"
readme.workspace = true
repository = "https://github.com/RibirX/Ribir/ribir"
version.workspace = true

[dependencies]
lyon_algorithms.workspace = true
lyon_path.workspace = true
ribir_core = {path = "../core", version = "0.0.0", features = ["png"]}
ribir_gpu = {path = "../gpu", version = "0.0.0"}
ribir_widgets = {path = "../widgets", version = "0.0.0"}

[dev-dependencies]
env_logger.workspace = true
png.workspace = true

[features]
default = ["wgpu_gl"]
wgpu_gl = ["ribir_gpu/wgpu_gl"]
