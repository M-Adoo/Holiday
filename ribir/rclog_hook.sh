#!/bin/bash

if [ "$MERGE_CHANGELOG" = true ]; then
  echo "$(rclog -t $NEW_VERSION -p ./CHANGELOG.md merge)" >| ./CHANGELOG.md
fi
