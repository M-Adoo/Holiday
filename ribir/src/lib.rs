pub use ribir_core as core;
pub use ribir_widgets as widgets;
pub mod prelude {
  pub use ribir_core::prelude::*;
  pub use ribir_widgets::prelude::*;
  pub mod app {

    #[cfg(feature = "wgpu_gl")]
    pub fn run(root: super::Widget) {
      let mut app = super::Application::default();
      let wnd = app.new_window(|native_wnd, ctx| {
        let size = native_wnd.inner_size();
        let p_backend = super::AppContext::wait_future(gpu::wgpu_backend_with_wnd(
          &native_wnd,
          super::DeviceSize::new(size.width, size.height),
          None,
          None,
          ctx.shaper.clone(),
        ));
        super::Window::new(native_wnd, p_backend, root.into_widget(), ctx)
      });
      app.exec(wnd);
    }
  }
}

use prelude::*;
#[cfg(feature = "wgpu_gl")]
pub fn wgpu_headless_wnd(root: Widget, ctx: AppContext, size: DeviceSize) -> Window {
  let p_backend = AppContext::wait_future(gpu::wgpu_backend_headless(
    size,
    None,
    None,
    ctx.shaper.clone(),
  ));
  Window::new(
    ribir_core::window::MockRawWindow {
      size: Size::from_untyped(size.to_f32().to_untyped()),
      ..Default::default()
    },
    p_backend,
    root,
    ctx,
  )
}
