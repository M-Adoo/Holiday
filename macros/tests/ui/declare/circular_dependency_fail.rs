use ribir::prelude::*;

fn main() {
  let _circular_dependency_err = widget! {
    Flex {
      SizedBox {
        id: id1,
        size: id2.size,
      }
      SizedBox {
        id: id2,
        size: id3.size,
      }
      SizedBox {
        id: id3,
        size: id1.size,
      }
    }
  };

  let _wrap_widget_circular_err = widget! {
    SizedBox {
      id: parent,
      size: Size::zero(),
      margin: child.margin.clone(),
      SizedBox{
        id: child,
        size: Size::zero(),
        margin: parent.margin.clone(),
      }
    }
  };

  let _data_flow_circular_err = widget! {
    SizedBox {
      id: a,
      size: Size::zero(),
    }
    modify_on a.size ~> a.size
  };

  let _data_flow_circular_field_skip_nc_pass = widget! {
    SizedBox {
      id: a,
      size: Size::zero(),
      SizedBox {
        id: b,
        size: a.size,
      }
    }
    change_on a.size ~> b.size
  };

  let _circular_follows_with_skip_nc_pass = widget! {
    SizedBox {
      id: a,
      size: Size::zero(),
    }
    change_on a.size ~> a.size
  };
}
