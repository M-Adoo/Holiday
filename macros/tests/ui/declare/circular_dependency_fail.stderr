error: There is a directly circle depends exist, this will cause infinite loop: id3.size ~> id1 , id2.size ~> id3 , id1.size ~> id2
  --> tests/ui/declare/circular_dependency_fail.rs:7:13
   |
7  |         id: id1,
   |             ^^^
8  |         size: id2.size,
   |         ^^^^  ^^^
...
11 |         id: id2,
   |             ^^^
12 |         size: id3.size,
   |         ^^^^  ^^^
...
15 |         id: id3,
   |             ^^^
16 |         size: id1.size,
   |         ^^^^  ^^^
   |
note: You can use `change` event to break circle, which will trigger only if the value really changed by compare if the value equal before modify and after.
  --> tests/ui/declare/circular_dependency_fail.rs:8:9
   |
8  |         size: id2.size,
   |         ^^^^
...
12 |         size: id3.size,
   |         ^^^^
...
16 |         size: id1.size,
   |         ^^^^

error: There is a directly circle depends exist, this will cause infinite loop: parent.margin ~> child , child.margin ~> parent
  --> tests/ui/declare/circular_dependency_fail.rs:23:11
   |
23 |       id: parent,
   |           ^^^^^^
24 |       size: Size::zero(),
25 |       margin: child.margin.clone(),
   |       ^^^^^^  ^^^^^^^^^^^^
26 |       SizedBox{
27 |         id: child,
   |             ^^^^^
28 |         size: Size::zero(),
29 |         margin: parent.margin.clone(),
   |         ^^^^^^  ^^^^^^^^^^^^^
   |
note: You can use `change` event to break circle, which will trigger only if the value really changed by compare if the value equal before modify and after.
  --> tests/ui/declare/circular_dependency_fail.rs:25:7
   |
25 |       margin: child.margin.clone(),
   |       ^^^^^^
...
29 |         margin: parent.margin.clone(),
   |         ^^^^^^

error: There is a directly circle depends exist, this will cause infinite loop: a ~> a
  --> tests/ui/declare/circular_dependency_fail.rs:36:11
   |
36 |       id: a,
   |           ^
...
39 |     modify_on a.size ~> a.size
   |                         ^
   |
note: You can use `change` event to break circle, which will trigger only if the value really changed by compare if the value equal before modify and after.
  --> tests/ui/declare/circular_dependency_fail.rs:36:11
   |
36 |         id: a,
   |  ___________^
37 | |       size: Size::zero(),
38 | |     }
39 | |     modify_on a.size ~> a.size
   | |_________________________^
